specs/makeStyles.md

```diff
@@ -0,0 +1,322 @@
+The `styles` passed to `makeStyles([[matcher, styles]])` call can be either an object or a function. The function is called with all theme tokens.
+Those tokens are resolved to CSS variable usages in build time. `ThemeProvider` component is responsible for setting the CSS variables in DOM and changing them when theme changes. When theme is switched, only the variables are changed, all styles remain the same.
+
+::Risk:: What if we will not be able to describe the theme changes using just different values?
+::Risk:: A theme consist of hundreds or even thousands of tokens. We need to verify browsers can handle this amount of CSS variables.
+
+#### IE11 compatibility
+
+IE11 does not support CSS variables. If we decide to support IE11, we will split the styles to the items which use tokens and those which do not. The ones which do not use tokens will be processed build time, for the ones with tokens, we will fall back to runtime evaluation.
+
+### Hash-based classnames
+
+We will use a hash (MurmurHash2) of the CSS property name and the value to generate a classname. That results in deterministic (=stable) classnames. With that we can support server side rendering without issues with hydration (#3692).
+
+### Passing overrides to children using `className`
+
+To avoid the ambivalence of using `className` vs `styles` props when passing overrides to children based on their type, we will use the dictionary side effect described in `useCSS` approach above.
+Parent component passes its overrides to a child as a list of classes in `className` prop. The child, takes the input classes and references the global dictionary to get back the necessary CSS attribute names to be able to perform the merge - the merge is implemented in `ax()` function as described in #16411.
+
+#### Alternative 1: encode the CSS property name to the hash
+
+Instead of using the global dictionary side effect, we can use part of the hash in the classname to encode the CSS property:
+![](makeStyles-classname.png)
+
+Then when merging in a child, we can merge the classes by replacing previous classnames with the same CSS property hash.
+This approach makes the classnames longer and according to the test does not make the merging any faster.
+
+#### Alternative 2: use special `styles` prop for overrides
+
+If the global dictionary does not perform as expected, we might return back to the previous v0 approach where we use a dedicated `styles` prop to pass overrides from parent to child.
+
+### Another libraries using similar approach
+
+This has been inspired by [Facebook stylex talk](https://www.youtube.com/watch?v=9JZHodNR184).
+Similar approach is also used in Attlasian [compiled v5](https://github.com/atlassian-labs/compiled).
+
+## Sample Code
+
+```jsx
+// Button.tsx
+const useButtonStyles = makeStyles([
+  [null, { cursor: 'pointer' }], // always applied
+  [props => props.primary, tokens => ({ backgroundColor: tokens.colors.primary.background })],
```

Comment:
**q:** 

- who injects `tokens` ?
```
[..., tokens => ({ backgroundColor: tokens.colors.primary.background })]
```

- with this API, how can one use dynamic props values within style definition?

```
// contrived example
[props => props.type === 'primary', {color: tokens.colors[props.type].background }]
```
 Is this designed like this on purpose ? So if we want different styling, we need to add another tuple to the array resolution ? (increasing codes size)

Reply:
> Is this designed like this on purpose ? So if we want different styling, we need to add another tuple to the array resolution ? (increasing codes size)

Purpose of this design is to produce the same classes always, in this case we can make there operations build time. If there is anything conditional on right, build time optimizations are no longer possible:

```jsx
{color: 'red' }, // will produce an atomic class
{ color: tokens.colors.primary.background }, // tokens will become CSS variables, again can become an atomic class 
{ color: tokens.color[props.type].background }, // ?! 💣 
```

Reply:
> who injects tokens ?

`tokens` or actually `theme` object is injected by `ThemeProvider` (`react-theme-provider`) or `FluentProvider` (`react-provider`). `react-make-styles` consumes required contexts and wires `makeStyles()` with proper `tokens` and RTL settings.

Reply:
**another q:**
- so dynamic properties in style definition part:
   -  will be solely injected from Context ?
   -  to define dynamic styles one has to use Callback api (with aforementioned data injection)
- how will type checking work ?
   - do I need to explicitly provide generic type ? `makeStyles<Props,Context>` ? or it's gonna be defined (for context) on global provider level with ability to extend on hierarchical providers ?  


Can we add this info the the spec as well please (typescript is important part of our API's) ?

Reply:
> If there is anything conditional on right, build time optimizations are no longer possible

That's true, but there are cases where you need to map from a prop to a shared color - there are currently 49 shared colors, that would mean you need to implement, maintain (and always execute) 49 different matchers. Let's try to thing if there is anyway how to handle this - programatically setting a css variable for component token as inline style for example?

cc @assuncaocharles as he has already hit this wall in `Badge`.

Reply:
> - will be solely injected from Context ?

As mentioned before, we will transform tokens usage to CSS variables for browsers that support this (this also allows to perform build time optimizations):

```jsx
const useStyles = makeStyles([
  [
    null,
    theme => ({
      backgroundColor: theme.alias.color.neutral.neutralBackground1,
      color: theme.alias.color.neutral.neutralForeground1
    })
  ]
]);
```

Will result in:

```css
.f16th3vw {
  color: var(--alias-color-neutral-neutralForeground1);
}
.fpnz0vo0 {
  background-color: var(--alias-color-neutral-neutralBackground1);
}
```

Currently, [`ThemeProvider`](https://github.com/microsoft/fluentui/tree/master/packages/react-theme-provider) injects both things: an object to React Context and CSS variables.

> - to define dynamic styles one has to use Callback api (with aforementioned data injection)

Currently, there are no dynamic styles, you can only use tokens from a theme in the same way as CSS variables:

```tsx
const useStyles = makeStyles([
  [
    null,
    theme => ({
      color: theme.alias.color.neutral.neutralForeground1,
      // effect is the same, but there are no type-checks and it work not in IE11
      color: "var(--alias-color-neutral-neutralForeground1)"
    })
  ]
]);
```

Reply:
> how will type checking work ?
> do I need to explicitly provide generic type ? makeStyles<Props,Context> ? or it's gonna be defined (for context) on global provider level with ability to extend on hierarchical providers ?

[`react-make-styles`](https://github.com/microsoft/fluentui/tree/master/packages/react-make-styles) version has only single generic type, `Tokens` are hard-coded to be a `Theme` from `react-theme`. Types for `Selectors` in `makeStyles<Selectors>` are necessary.

Reply:
https://github.com/microsoft/fluentui/pull/16793 has an example of converged styles, it might clarify usage. I appreciate any suggestions how we can improve current document 👍

Reply:
This is in the spec:

> ### Applying theme to styles
> No matter what theme is used, the component styles are always the same. The only way how to change the component appearance is through tokens which can be used in style values.
> The `styles` passed to `makeStyles([[matcher, styles]])` call can be either an object or a function. The function is called with all theme tokens.
> Those tokens are resolved to CSS variable usages in build time. `ThemeProvider` component is responsible for setting the CSS variables in DOM and changing them when theme changes. When theme is switched, only the variables are changed, all styles remain the same.

@Hotell do you think it is sufficient or is there still anything what should be added to the spec?

Reply:
Overall I'd be great to add some graphics or more code examples to make it easier to grok.

From my experience when using types for implementation, it should be part of any spec to showcase the interfaces ( which again makes it more "obvious" ).

**For example (simplified example - no need to go deeper imho):**

```
// type annotation of proposed API
function makeStyles<Props>(cssDefinitions: Array<[Matchers<Props>,Styles]>): Hook
// where
type Hook = (props:Props)=>string
type Props = Record<...>
type Matchers<P> = (props:P)=>boolean
type Styles = (theme:Theme)=>CssDefinition
// where
// global tokens/theme definition provided in ThemeProvider(React Context)
type Theme = {}
``` 

If I saw that I would have not so many questions for sure :D 

Appreciate your answers though 👍